pg_search/src/postgres/storage/linked_items.rs

```diff
@@ -346,6 +357,158 @@ impl<T: From<PgItem> + Into<PgItem> + Debug + Clone + MVCCEntry> LinkedItemList<
+        // ... but not both.
+        let mut blockno = original_header_lock
+            .page()
+            .contents::<LinkedListData>()
+            .start_blockno;
+        assert!(
+            blockno != pg_sys::InvalidBlockNumber,
+            "Must contain at least one block."
+        );
+        let mut previous_new_buffer: Option<BufferMut> = None;
+        while blockno != pg_sys::InvalidBlockNumber {
+            let buffer = self.bman.get_buffer(blockno);
+            let page = buffer.page();
+            let mut offsetno = pg_sys::FirstOffsetNumber;
+            let max_offset = page.max_offset_number();
+
+            let mut new_buffer = cloned.bman.new_buffer();
+            let new_blockno = new_buffer.number();
+            let mut new_page = new_buffer.init_page();
+
+            // Link the new block in with the previous block, or with the header.
+            if let Some(mut previous_new_buffer) = previous_new_buffer {
+                previous_new_buffer
+                    .page_mut()
+                    .special_mut::<BM25PageSpecialData>()
+                    .next_blockno = new_blockno;
+            } else {
+                header_buffer
+                    .page_mut()
+                    .contents_mut::<LinkedListData>()
+                    .start_blockno = new_blockno;
+            }
+
+            while offsetno <= max_offset {
+                if let Some((deserialized, _)) = page.read_item::<T>(offsetno) {
+                    let PgItem(item, size) = deserialized.into();
+                    let new_offsetno = new_page.append_item(item, size, 0);
+                    assert!(new_offsetno != pg_sys::InvalidOffsetNumber);
+                }
+                offsetno += 1;
+            }
+
+            *new_page.special_mut::<BM25PageSpecialData>() =
+                (*page.special::<BM25PageSpecialData>()).clone();
+            blockno = page.next_blockno();
+            previous_new_buffer = Some(new_buffer);
+        }
+
+        AtomicGuard {
+            original: self,
+            original_header_lock,
+            cloned: Some(cloned),
+        }
+    }
+}
+
+pub struct AtomicGuard<'s, T: From<PgItem> + Into<PgItem> + Debug + Clone + MVCCEntry> {
+    original: &'s mut LinkedItemList<T>,
+    original_header_lock: BufferMut,
+    cloned: Option<LinkedItemList<T>>,
+}
+
+impl<T: From<PgItem> + Into<PgItem> + Debug + Clone + MVCCEntry> Deref for AtomicGuard<'_, T> {
+    type Target = LinkedItemList<T>;
+
+    fn deref(&self) -> &Self::Target {
+        self.cloned.as_ref().expect("Cannot deref after commit!")
+    }
+}
+
+impl<T: From<PgItem> + Into<PgItem> + Debug + Clone + MVCCEntry> DerefMut for AtomicGuard<'_, T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.cloned
+            .as_mut()
+            .expect("Cannot deref_mut after commit!")
+    }
+}
+
+impl<T: From<PgItem> + Into<PgItem> + Debug + Clone + MVCCEntry> AtomicGuard<'_, T> {
+    pub fn commit(&mut self) {
```

Comment:
should this instead be `pub fn commit(mut self)`?

Then I think the internal `cloned` field could just be a `LinkedItemList<T>`.  Combined, this would eliminate the possibility of using `AtomicGuard` incorrectly.